Update read_watersurfaces for use with v 1.1

cecileherr opened the pull request:
**Changes:**
- if extended == TRUE: add a variable `connectivity_name` with a description of the connectivity to streams
- update documentation to version 1.1
- support version 1.1
=> default: works with version 1.1 
=> working with version 1.0 is still possible 

**Note about the different versions:**
- I isolated a few v1.0-specific `mutate` statements in an `if-else statement` - to make it easy to modify the code in the future-, but actually these `mutate` for v1.0 would be perfectly compatible with v1.1
- v1.1 is a geopackage and v1.0 a shapefile. The user needs to specify the `version` argument to be able to read version 1.0 (if not: generates a meaningful error)
- the user can specify version 1.0 as argument without needing to change the file argument
- if a custom path is used (`file `argument), the file will be treated as version 1.1 by default without any warning (unless the user specifies the `version` argument)

There are several aspects that could be better managed (e.g. possible conflicts between `file` and `version` arguments) but I think it might be more interesting to work directly towards a more [structured implementation](https://inbo.github.io/n2khab/index.html#intention-for-the-future) of version control (with md5, ...). 
Any thoughts on this?

florisvdh commented:
> I think it might be more interesting to work directly towards a more [structured implementation](https://inbo.github.io/n2khab/index.html#intention-for-the-future) of version control (with md5, ...). 

Agreed @cecileherr. If I'm to implement that (current plan) I need to postpone that till later this year and in the meantime quick-fixes are needed. The new function behaviour as you describe in your comment sounds very good (will review soon). But yes, we definitely need an overarching system which can tackle all this version/file checking (I have an idea), especially for extra versions in the future. Also involves #113.

If you're up to do the coding yourself sooner: welcome - then we best discuss on the ideas in person first. Still, we will always need version-specific code inside the functions (file contents differ per version), at least for all versions which we continue supporting (which I'd like to keep up at least as long as the approach to handle versions is evolving).

florisvdh reviewed (commented):
Thank you @cecileherr! I like the user-friendly spirit for the filepath / version-dependency.

- > I isolated a few v1.0-specific `mutate` statements in an `if-else statement` - to make it easy to modify the code in the future-, but actually these `mutate` for v1.0 would be perfectly compatible with v1.1

  :+1: it is the best way IMO, version-specific code is best not run unnecessarily

- > the user can specify version 1.0 as argument without needing to change the file argument

  :+1: 

- main consideration for v1.1 is probably the different layers present in the GeoPackage; I would tend to do something with it (i.e. return/join if useful), see also https://github.com/inbo/n2khab-preprocessing/pull/53#pullrequestreview-604942169. If it is deliberately not used, we should at least say something about it in the _Details_ section.

- regarding connectivity: indeed the function works for version 1.0 as well:

  ```r
  > ws1 <- read_watersurfaces(version = "watersurfaces_v1.0", extended = TRUE)
  > ws1 %>% st_drop_geometry %>% dplyr::count(connectivity, connectivity_name)
  # A tibble: 3 x 3
    connectivity              connectivity_name                    n
    <fct>                     <fct>                            <int>
  1 DoorstromingMetPeilbeheer DoorstromingMetPeilbeheer            1
  2 geïsoleerd                niet verbonden met een waterloop     2
  3 NA                        NA                               86023
  ```
  However, given previous point (possibility to extract values from table in v1.1) adding connectivity names in v1.0 could be weighed against the very sparse data on connectivity in v1.0 (still not much in v1.1 though), to decide whether to hardcode translations or only translate for v1.1, using the included table in the latter case. I'd like to await your further findings in inbo/n2khab-preprocessing#53, on the usability of the non-spatial tables, so that we can decide this better. Anyway your proposed approach does work well!

- _Usage_ section (documentation):

  I do like the idea to set `file`, however I see one current drawback and it is the _Usage_ section in the documentation. In general, it makes clear which specific file is taken by default, unless it is a bunch of files (in which case a folder is shown). Showing the folder in _Usage_ while just one *.gpkg file is taken by default, breaks this idea.

  What do you think about this alternative:

  - setting the gpkg path as default in `function(...)` (hence, show in _Usage_)
  - then setting the v1.0 folder only `if (version == "watersurfaces_v1.0" && missing(file))`
  - relating to "_the user can specify version 1.0 as argument without needing to change the file argument_", add this in the _Details_.

Do know I'm pleased we mainly have minor (aesthetic) things to discuss :smile_cat:. You pay much attention to things already.

Regarding commit messages:

- they are very good, thanks for adoping my preferred `subject: do this` style, I find that very convenient when searching/browsing through a repo's log.
- advice: make use of the message body if otherwise you obtain a long title (cf. ones that drop off the line in GitHub). The body is separated from the title with a whiteline; it can have multiple paragraphs (whitelines) if you want to add more explanation etc. I use an asterisk in the title to mark that a body is present, to make clear in logs that some commit messages provide more explanation. I.e. `subject: do this *`. Specifically for GitHub this is not needed (marked with `...`), but in general (cf `git log --oneline` and so on) it's useful.

florisvdh commented:
>   What do you think about this alternative:
>
>  - setting the gpkg path as default in `function(...)` (hence, show in _Usage_)
>  - then setting the v1.0 folder only `if (version == "watersurfaces_v1.0" && missing(file))`
>  - relating to "_the user can specify version 1.0 as argument without needing to change the file argument_", add this in the _Details_.

@cecileherr I think the following will be better and more future-proof:

- no explicit default for `file` while still showing in `Usage` that it is _optional_, with `function(file = NULL, )`. Cf. https://adv-r.hadley.nz/functions.html#missing-arguments
  - This may be a more robust way for the future (less maintenance needed), considering that we'll use metadata tables already inside the package that will tell where files are by default (for each data source, per version). (And which can be returned as well)
- `if (missing(file))`, i.e. user did not actively specify `file`, set the file for each version
- change the `file` argument documentation to state more clearly that function will use defaults based on `version`. I.e. override the default with an explicit `@param file ...`

What do you think?

florisvdh commented:
@cecileherr This idea comes a bit late to the party, but for the `é` and `ï` cases, can you check whether the following trick, building on the results of inbo/n2khab-preprocessing#53, works as well? It would make translating the encoding from UTF-8 to latin1 (and actively defining as latin1) unnecessary, which may not be an ideal solution IMO. If the following works, it would be a better cross-platform solution.

```r
Encoding(x) # I think this returns "unknown"
Encoding(x) <- "UTF-8"
print(x) # to check the effect: are characters printed correctly?
```

If I'm correct you will see that, originally, R does not know about the encoding of the data (`Encoding(x)` returns `"unknown"`). Hence it assumes the native encoding of the operating system. Which, in the case of GeoPackages (= UTF-8 or UTF-16), may always be wrong in Windows (unless GDAL attempts to take care of it, don't know for sure - and maybe `sf` plays a role in it).

The above sets the correct encoding for the characters at hand, and maybe that already suffices. But I'm not sure; I think I've tried it before, maybe not (always?) successfully.

florisvdh commented:
For the character encoding, some ideas along the same line can be found in r-spatial/sf#5. A good workflow might also be achieved by using `rgdal` to set `encoding = "UTF-8"` and then converting the object to `sf`:

- > Note that `rgdal::readOGR()` has `encoding=` and `use_iconv=` arguments
- > you could either use `rgdal::readOGR()` as suggested by rsbivand and convert that to an sf object with `st_as_sf()`.

If you'd follow such route, then don't Import `rgdal` but add it to Suggests (file `DESCRIPTION`), and look at how suggested packages are used in `n2khab` functions. However this route will slow down the usage of the function.

cecileherr commented:
About the **encoding problems**:

> Encoding(x) <- "UTF-8"

Implemented in 04e0d83 for the non geometrical layers (for an unknown reason the characters were correctly recognized in the geometrical layer of the Geopackage). 
The encoding for the non geometrical layers was indeed unknown. If I set it with `Encoding() <-`, it becomes UTF-8 but only for the elements with special characters, the encoding of the other elements stays "unknown".

 ```
 connectivitytransl <- read_sf(file, layer = "LktCONNECT") %>%
                    # mutate(across(where(is.character),
                    #               .fns = function(x){return(`Encoding<-`(x, "UTF-8"))})) %>%
                    mutate_if(., is.character,
                              .funs = function(x){return(`Encoding<-`(x, "UTF-8"))}) 

Encoding(connectivitytransl$Code)
#[1] "UTF-8"   "unknown" "unknown"
Encoding(connectivitytransl$Omschr)
#[1] "unknown" "UTF-8"   "UTF-8" 
```
I guess it should be ok.

Side note about the commented part: I know `mutate_if` is deprecated, but I decided against using the `across` alternative because of this: (`where` not exported) https://github.com/r-lib/tidyselect/issues/201

cecileherr commented:
> > What do you think about this alternative:
> > 
> > * setting the gpkg path as default in `function(...)` (hence, show in _Usage_)
> > * then setting the v1.0 folder only `if (version == "watersurfaces_v1.0" && missing(file))`
> > * relating to "_the user can specify version 1.0 as argument without needing to change the file argument_", add this in the _Details_.
> 
> @cecileherr I think the following will be better and more future-proof:
> 
> * no explicit default for `file` while still showing in _Usage_ that it is _optional_, with `function(file = NULL, )`. Cf. https://adv-r.hadley.nz/functions.html#missing-arguments
>   
>   * This may be a more robust way for the future (less maintenance needed), considering that we'll use metadata tables already inside the package that will tell where files are by default (for each data source, per version). (And which can be returned as well)
> * `if (missing(file))`, i.e. user did not actively specify `file`, set the file for each version
> * change the `file` argument documentation to state more clearly that function will use defaults based on `version`. I.e. override the default with an explicit `@param file ...`
> 
> What do you think?

See 6e12e73 (and documentation in d26d5b0 and 4491ef7)
- Tried the second option (set `file = NULL` )
- Added a few `ifs` to retrieve an error when the `file` and the `version` are conflicting

cecileherr commented:
Many thanks for the detailed review (as usual ;-) ) and tips!

Pushed a **few commits** to cover your suggestions:
- declared all named objets with `.data$` 4e60448
- changed `is.null(file)` to `missing(file)` and adapted the documentation accordingly ba8c91f & dbbc07a
- added `str_replace` in the `@importFrom` (I thought I had checked this a 100 times :D) 4e60448
- cosmetic changes  999d75c

About **.data$**

> Best check this note disappears locally by running the check in the RStudio Build pane (just pressing ctrl+shift+E should do the job).

Unfortunately it does not work on my computer yet (seems I have to install extra software and I need admin rights, so no immediate fix for this. I hope you don't mind if I keep polluting the online repository with my failing checks while I try to solve the issue locally)

About **NAMESPACE**

I am a bit perplexed about this one. As far as I can remember I only ran `roxygen2::roxygenise() ` locally, I don't think I changed the file directly (or it was not on purpose). Anyway probably better practice to run `devtools::document()`

About whether to include full names for **usage and depth class**?
- depth class: I also hesitated and came to the same conclusion as you: still too many corrections needed in the data source, beter to add the translation in a next version
- usage: only a few categories present in the dataset at the moment and their meaning is quite clear e.g. natuur, tuin_park, waterberging... I would add the translation in a next version if needed, but if you think it can already be useful for v1.1: just let me know.

cecileherr commented on R/read_habitatdata.R:
```diff
@@ -530,13 +571,15 @@ read_watersurfaces <-
-                wfd_type = .data$wfd_type %>%
+            mutate(depth_class = str_replace(string = depth_class,
+                                             pattern = "\u2265",
+                                             replacement = ">=")) %>%
```
see 4e60448

cecileherr commented on R/read_habitatdata.R:
```diff
@@ -458,66 +476,89 @@ read_watersurfaces_hab <-
-        if (file == file.path(fileman_up("n2khab_data"),
-                              "10_raw/watersurfaces")) {
-
+        if (is.null(file)) {
```
I followed your suggestion, see ba8c91f

florisvdh reviewed (approved):
Excellent @cecileherr, thanks for the follow-up. If there was nothing more that you planned to add, then feel free to merge.

The note about undefined global variables is gone :+1:.

> About NAMESPACE I am a bit perplexed about this one.

See 4103538.
I follow https://r-pkgs.org/man.html#man-workflow and https://r-pkgs.org/namespace.html#imports, but clearly the latter advice has evolved since I learned it - rather the default advice is now to always do `package::function()` instead of only doing that for packages in `Suggests:`. Hm, that's not as handy for the developer IMHO.

> whether to include full names for **usage and depth class**

I agree with you.

> seems I have to install extra software

I believe all Windows users should have RTools installed by default (if not: better let it happen), I don't know about other needs?

> I hope you don't mind if I keep polluting the online repository with my failing checks while I try to solve the issue locally

Sure, don't mind.

If you like to experiment more or feel limited to do that in the PR-context, you can always do it in a separate branch (outside PR): on pushing, it also triggers the R-CMD-check workflow, on the branch tip. Outside a PR, it's also more convenient to update/overwrite your git history (invoking new commit hashes) and force-push your branch untill all is fine. After that you can merge your branch locally (with main feature branch checked out: `git merge experimentalbranch`), and delete `experimentalbranch` on the remote (`git push --delete origin experimentalbranch`). I could demo that if you like.

Also, on GitHub you're free to 'resolve' comments that you consider resolved for yourself (if someone else doesn't agree, (s)he can always unresolve or add another reply). It could make it visually easier for you to follow up (resolving makes them hidden). You choose, though.

cecileherr commented:
> I follow https://r-pkgs.org/man.html#man-workflow and https://r-pkgs.org/namespace.html#imports, but clearly the latter advice has evolved since I learned it - rather the default advice is now to always do package::function() instead of only doing that for packages in Suggests:. Hm, that's not as handy for the developer IMHO.

Ok, just let me known if we should start using package::function() systematically from then on! (on the other hand the article states that it is a bit slower to use package::function() than @importFrom pkg fun +function() )

About the **local CMD check**: for now I need to install qpdf and change the system path in Win. 

> If you like to experiment more or feel limited to do that in the PR-context, you can always do it in a separate branch (outside PR): on pushing, it also triggers the R-CMD-check workflow, on the branch tip. Outside a PR, it's also more convenient to update/overwrite your git history (invoking new commit hashes) and force-push your branch untill all is fine. After that you can merge your branch locally (with main feature branch checked out: git merge experimentalbranch), push it, and delete experimentalbranch on the remote (git push --delete origin experimentalbranch). I could demo that if you like.

Yes, might be the way to go. Maybe next time, when I get many many reactions to a PR ;-)

As far as I can see, the failed check for macOS is not related to n2khab (?) so I suppose I can merge this PR.

florisvdh commented:
> the failed check for macOS is not related to n2khab (?) so I suppose I can merge this PR.

Yes, see #120. I'll merge your PR into `dev-0.5.0`!

florisvdh commented:
> let me known if we should start using package::function() systematically

I'd rather not change current approaches. `@importFrom` is not going to disappear; it's a matter of philosophy. I prefer one coding strategy within the same package - sometimes it can mean to migrate everything.

florisvdh commented:
@hansvancalster @ElsLommelen does this sound familiar; any recommendations about how @cecileherr can solve below problem with locally running `R CMD check`? Double quoted is from me, single quoted from her.

>>    Best check this note disappears locally by running the check in the RStudio Build pane (just pressing ctrl+shift+E should do the job).
>
>Unfortunately it does not work on my computer yet (seems I have to install extra software and I need admin rights, so no immediate fix for this.

>>I believe all Windows users should have RTools installed by default (if not: better let it happen), I don't know about other needs?
>
> About the local CMD check: for now I need to install qpdf and change the system path in Win.

hansvancalster commented:
> Unfortunately it does not work

@cecileherr can you show us the actual error? Does it say which extra software is needed?

cecileherr commented:
> > Unfortunately it does not work
> 
> @cecileherr can you show us the actual error? Does it say which extra software is needed?

qpdf is missing (the error message was pretty clear, I can add it here later if it can be useful), see also https://stackoverflow.com/questions/15035150/r-cmd-check-as-cran-warning 
It is probably not hard to solve, but I need IT to modify the Win system path (admin rights required)

hansvancalster commented:
@cecileherr yes indeed, you need IT to solve that. We should probably also document this in admin installation instructions (it affects all Windows 10 users who want to use `devtools::check()`).
